qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1963 @@
+            pulse.u3(0, math.pi, 0, 0)
+            pulse.cx(0, 1)
+
+
+            # It is also be possible to call a preexisting schedule
+            tmp_sched = pulse.Schedule()
+            tmp_sched += pulse.Play(gaussian_pulse, d0)
+            pulse.call(tmp_sched)
+
+            # We also support:
+
+            # frequency instructions
+            pulse.set_frequency(5.0e9, d0)
+
+            # phase instructions
+            pulse.shift_phase(0.1, d0)
+
+            # offset contexts
+            with pulse.phase_offset(math.pi, d0):
+                pulse.play(gaussian_pulse, d0)
+
+The above is just a small taste of what is possible with the builder. See the
+rest of the module documentation for more information on its
+capabilities.
+
+.. warning::
+    The pulse builder interface is still in development and is
+    subject to change.
+"""
+import collections
+import contextvars
+import functools
+import itertools
+from contextlib import contextmanager
+from typing import (
+    Any,
+    Callable,
+    ContextManager,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Set,
+    Tuple,
+    TypeVar,
+    Union
+    )
+
+import numpy as np
+
+from qiskit import circuit
+from qiskit.circuit.library import standard_gates as gates
+from qiskit.pulse import channels
+from qiskit.pulse import configuration
+from qiskit.pulse import exceptions
+from qiskit.pulse import instructions
+from qiskit.pulse import macros
+from qiskit.pulse import pulse_lib
+from qiskit.pulse import transforms
+from qiskit.pulse.instructions import directives
+from qiskit.pulse.schedule import Schedule
+
+#: contextvars.ContextVar[BuilderContext]: active builder
+BUILDER_CONTEXTVAR = contextvars.ContextVar("backend")
+
+T = TypeVar('T')  # pylint: disable=invalid-name
+
+
+def _compile_lazy_circuit_before(function: Callable[..., T]
+                                 ) -> Callable[..., T]:
+    """Decorator thats schedules and calls the lazily compiled circuit before
+    executing the decorated builder method."""
+    @functools.wraps(function)
+    def wrapper(self, *args, **kwargs):
+        self._compile_lazy_circuit()
+        return function(self, *args, **kwargs)
+    return wrapper
+
+
+class BackendNotSet(exceptions.PulseError):
+    """Raised if the builder context does not have a backend."""
+
+
+def _requires_backend(function: Callable[..., T]) -> Callable[..., T]:
+    """Decorator a function to raise if it is called without a builder with a
+    set backend.
+    """
+    @functools.wraps(function)
+    def wrapper(self, *args, **kwargs):
+        if self.backend is None:
+            raise BackendNotSet(
+                'This function requires the builder to '
+                'have a "backend" set.')
+        return function(self, *args, **kwargs)
+    return wrapper
+
+
+class _PulseBuilder():
+    """Builder context class."""
+
+    def __init__(self,
+                 backend=None,
```

Comment:
There's a circular dependency between the backends and the builder? We should remove any dependency within the backend on the builder

Reply:
Could you help point out the particular circular dependency?

Reply:
there isn't? ok then:
```suggestion
                 backend: BaseBackend = None,
```